include/seqan3/search/dream_index/binning_directory.hpp

```diff
@@ -0,0 +1,218 @@
+     * If you want to add more bins while keeping the size constant, you need to rebuild the binning_directory.
+     */
+    inline void resize(size_t const new_bins)
+    {
+        if (new_bins < bins)
+        {
+            throw std::invalid_argument{"The new number of new bins must be greater"
+                                        " or equal to the current number of bins."};
+        }
+
+        size_t bin_width_ = (new_bins + 63) >> 6;
+
+        if (bin_width_ == bin_width)
+        {
+            bins = new_bins;
+            return;
+        }
+
+        size_t block_size_ = bin_width_ << 6;
+        size_t bits_       = block_count * block_size_;
+
+        size_t idx_{bits_}, idx{data.size()};
+        size_t delta = block_size_ - block_size + 64;
+
+        data.resize(bits_);
+
+        for (size_t i = idx_, j = idx; j > 0; i -= block_size_, j -= block_size)
+        {
+            size_t stop = i - block_size_;
+
+            for (size_t ii = i - delta, jj = j - 64; stop && ii >= stop; ii -= 64, jj -= 64)
+            {
+                uint64_t old = data.get_int(jj);
+                data.set_int(jj, 0);
+                data.set_int(ii, old);
+            }
+        }
+
+        bins = new_bins;
+        bin_width = bin_width_;
+        block_size = block_size_;
+    }
+};
+
+/*!\brief A binning directory using a compressed bitvector.
+ * \implements seqan3::BinningDirectory
+ *
+ * \details
+ *
+ * This datastructure represents an immutable binning_directory, i.e. only querying is supported.
+ * The binning_directory_compressed can only be constructed from a binning_directory, in which case
+ * the underlying bitvector is compressed.
+ */
+class binning_directory_compressed : public detail::binning_directory_base<sdsl::sd_vector<>>
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    binning_directory_compressed() = default;                                                 //!< Defaulted.
+    binning_directory_compressed(binning_directory_compressed const &) = default;             //!< Defaulted.
+    binning_directory_compressed & operator=(binning_directory_compressed const &) = default; //!< Defaulted.
+    binning_directory_compressed(binning_directory_compressed &&) = default;                  //!< Defaulted.
+    binning_directory_compressed & operator=(binning_directory_compressed &&) = default;      //!< Defaulted.
+    ~binning_directory_compressed() = default;                                                //!< Defaulted.
+
+    /*!\brief Construct from seqan3::binning_directory.
+     * \param bd     The seqan3::binning_directory.
+     */
+    binning_directory_compressed(binning_directory const & bd)
+    {
+        std::tie(bins, bin_width, block_size, block_count, num_hash, pre_hash) =
+            std::tie(bd.bins, bd.bin_width, bd.block_size, bd.block_count, bd.num_hash, bd.pre_hash);
+
+        data = sdsl::sd_vector<>{bd.data};
```

Comment:
This probably involves copying or is there a constructor that takes a temporary and is more efficient? Then I would consider that the constructor only takes temporaries